### ISSUE TITLE
CPU capped at 100%, or hanging & capped at the same time.

### ISSUE THREAD

**Issue by robertpauson** [gakonst/ethers-rs#1896]
Used Chainstack, Quicknode, Infura. The code looks messy here and there but this is the only version able to reconnect my program that was also running AVAX settings, where the Ws errors were crazy sometimes on QuickNode.

```
#![allow(unused_assignments)]
#![allow(non_snake_case)]
use std::sync::{Arc, Mutex};
use std::collections::HashMap;
use tokio;
use tokio::time::{
    Duration,
    timeout
};
use ethers::prelude::*;

const NODEAP_BSC:&str = "";
const NODEAP_ETH:&str = "";
const NODEAP_POLY:&str = "";

const GET_BLOCK_TIMEOUT:u64 = 30;
const RECON_TIMEOUT:u64 = 30;


#[tokio::main]
async fn main(){
    let active_chains = vec!["BSC", "ETH", "POLY"];

    let mut PROVIDERZ:HashMap<String,Arc<Provider<Ws>>> = HashMap::new();
    let mut CHAINZ:HashMap<&str,&str> = HashMap::new();
    CHAINZ.insert("BSC", NODEAP_BSC);
    CHAINZ.insert("ETH", NODEAP_ETH);
    CHAINZ.insert("POLY", NODEAP_POLY);
    let CHAINZ = Arc::new(Mutex::new(CHAINZ));

    for name in &active_chains{
        let ap = CHAINZ.lock().unwrap()[name];
        let provider = match Provider::<Ws>::connect(ap).await{
            Ok(x)=>x,
            Err(x)=>{
                println!("[{}] can't connect: {:?}",&name, x);
                continue
            }
        };
        let provider: Arc<Provider<Ws>> = Arc::new(provider);
        PROVIDERZ.insert(name.to_string(),provider);
    };
    let PROVIDERZ = Arc::new(Mutex::new(PROVIDERZ));
    
    for name in active_chains{
        tokio::spawn({
            obtain_blocks(
                Arc::clone(&PROVIDERZ),
                Arc::clone(&CHAINZ),
                name.clone(),
            )
        });
    };
    loop{};
}

async fn obtain_blocks(
    PROVIDERZ:Arc<Mutex<HashMap<String,Arc<Provider<Ws>>>>>,
    CHAINZ:Arc<Mutex<HashMap<&str,&str>>>,
    chain:&str
){
    'ensure_connect: loop{
        let ap = CHAINZ.lock().unwrap()[&chain];
        let provider_: Provider<Ws> = match timeout(Duration::from_secs(30), Provider::<Ws>::connect(ap)).await{
            Ok(x)=>match x{
                Ok(x)=>x,
                Err(x)=>{
                    println!("[{:?}] err0: [{:?}]", &chain, x);
                    tokio::time::sleep(Duration::from_secs(RECON_TIMEOUT)).await;
                    continue 'ensure_connect
                }
            },
            Err(x)=>{
                println!("[{:?}] err1: [{:?}] -- timeout tokio V", &chain, x);
                tokio::time::sleep(Duration::from_secs(RECON_TIMEOUT)).await;
                continue 'ensure_connect
            }
        };
        PROVIDERZ.lock().unwrap().insert(chain.clone().to_owned(), Arc::clone(&Arc::new(provider_)));
        let provider = PROVIDERZ.lock().unwrap()[chain].clone();
        let mut stream = match timeout(Duration::from_secs(RECON_TIMEOUT), provider.watch_blocks()).await{
            Ok(x)=>{ match x{
                Ok(x)=>x,
                Err(x)=>{
                    println!("[{}] err2: {:?}",&chain, x);
                    tokio::time::sleep(Duration::from_secs(RECON_TIMEOUT)).await;
                    continue 'ensure_connect
                }
            }},
            Err(x)=>{
                println!("[{}] err3: {:?}",&chain, x);
                continue 'ensure_connect
            }
        };
        loop{
            while let Some(b_) = match timeout(Duration::from_secs(GET_BLOCK_TIMEOUT),stream.next()).await {Ok(x)=>x,Err(x)=>{println!("[{}] -stream err- {:?}",&chain, x);continue 'ensure_connect}} {
                let provider = PROVIDERZ.lock().unwrap()[chain].clone();
                let mut b=U64::from(0);
                'retry_call:loop{
                    let b_ = match timeout(Duration::from_secs(GET_BLOCK_TIMEOUT), provider.get_block(b_)).await{
                        Ok(x)=>match x{
                            Ok(x)=>x,
                            Err(x)=>{
                                println!("[{:?}] get_block error0: {:?}",&chain, x);
                                tokio::time::sleep(Duration::from_secs(GET_BLOCK_TIMEOUT)).await;
                                continue 'ensure_connect
                            }
                        },
                        Err(x)=>{
                            println!("[{:?}] get_block error1: {:?}",&chain, x);
                            tokio::time::sleep(Duration::from_secs(GET_BLOCK_TIMEOUT)).await;
                            continue 'ensure_connect
                        }
                    };
                    let b_ = match b_{
                        Some(x)=>x,
                        None=>{
                            println!("get_block error2");
                            tokio::time::sleep(Duration::from_secs(GET_BLOCK_TIMEOUT)).await;
                            continue 'retry_call
                        }
                    };
                    b = match b_.number{
                        Some(x)=>{x},
                        None=>{
                            println!("get_block error3");
                            tokio::time::sleep(Duration::from_secs(GET_BLOCK_TIMEOUT)).await;
                            continue 'retry_call
                        }
                    };
                    println!("[{}] [block: {:?}]",chain, &b);
                    break 'retry_call
                };
            }
        }
    }
}
```

This above minimal version gets stuck at 100% CPU. More bulky versions - which include AVAX, and multiple other functionalities completely hang, all threads that get blocks from stream, and threads /tasks that make requests through reqwest. I thought it was a reqwest bug initially, because last printed lines in most hangs were incompleted, forever-hanging reqwest calls. I was surprised to see it happened with the above as well.

**Comment by mattsse** [comment]
what tokio features are you using?

**Comment by robertpauson** [comment]
`tokio = { version ="1.18.2", features = ["full"] }`

**Comment by mattsse** [comment]
seeing this

```
 loop{};
```

would explain a 100% utilized core

and I assume there's a deadlock somewhere.
tokio tasks usually shouldn't be blocking

I don't think you even need the to guard the provider via a mutex, can just clone the entire thing after Arcing it?

**Comment by robertpauson** [comment]
>  loop{};

I've put it in repro only, because I was lazy, it hangs with joins as well which I normally use (or for unnamed task unjoined, can this cause it?).

>I don't think you even need to guard the provider via a mutex, can just clone the entire thing after Arcing it?

In different modules built on top of it, if something fails repeatedly, I reconnect and put back the reconnected Arc into the hashmap, not sure if I can do without, but I'll try it out thanks.

**Comment by mattsse** [comment]
you could try spawning blocking tasks instead first

**Comment by robertpauson** [comment]
how? I've searched through google and experimented and the threads when passing cloned Arcs (regular start giving errors about static borrow) the program compiled, then didn't get one block, I think the way I'm spawning threads won't work with blocking?

anyways, I've removed loop and using join! 
I've also written a restarter thread, that works 99% of the time

But the program still hanged once, and restarter still hanged. No errors in output. Weird.

**Comment by robertpauson** [comment]
IDK how to properly debug such locks, how to get full output/stacktrace of every thread

**Comment by mattsse** [comment]
I haven't looked at the code so idk what it does.

but there some await points so could just be that all of them aren't really that computationally demanding.

idk what you mean by hanging? 
did you remove the Mutex?

**Comment by robertpauson** [comment]
the script also hanged on an Ubuntu server, just checked. Cpu 400%. Has anyone had luck reproducing & fixing the hanging?

**Comment by dbelv** [comment]
@robertpauson 
So I was able to reproduce the issue with the code above.

![image](https://user-images.githubusercontent.com/90095005/210543957-41810560-44f1-491d-8cca-b409edf4907a.png)

`loop {}` is a form of busy waiting and you have async tasks so the main execution loop will never give up control once it gets control. This is what is causing the 100% CPU slam - the execution is looping forever, which is very different to doing nothing.

When forcing the main execution loop to give up control (and never get it back unless all tasks finish), with a dodgy task push into vec and await (please critically assess for your needs :))
```rust
    let mut v = Vec::new();
    for name in active_chains{
        v.push(tokio::spawn({
            obtain_blocks(
                Arc::clone(&PROVIDERZ),
                Arc::clone(&CHAINZ),
                name.clone(),
            )
        }));
    };

    for job in v {
        job.await.unwrap();
    }

    println!("All jobs finished!");
    loop{};
```

CPU usage is what you would expect (basically nothing):
![image](https://user-images.githubusercontent.com/90095005/210545001-be31dfc9-a372-47bb-88fa-59fe1c219002.png)

In regards to the streams stopping, I believe this is caused by the issue above for the reason that async tasks still execute concurrently (note not in parallel) so the main execution loop and any tasks *must* at some time give up control in order for the another task to "run" (it might not be on the same thread it started on but its running).  Giving up control is done with the `await`s however this doesn't mean the scheduler (the component telling which task to run) is doing something ideal for your scenario (like round robin). 

This is much harder to debug and confirm, however I would sugest running the section reading the stream on a separate thread (`thread::spawn`) so everything is running in parallel and then logging/printing when you handle a message, if you find everything working without any issues there is likely a task dominating the execution and you'll need to re-think your design.

**Comment by dbelv** [comment]
> how? I've searched through google and experimented and the threads when passing cloned Arcs (because tokio::spawn_blocking(Arc::clone(...)) was giving errors about static borrow so I used .clone() in previous scope and it worked) so the program compiled, then didn't get even one block

In regards to static boundary issues that's the way to fix them, clone outside the scope so the scope can take ownership of a variable passed, otherwise it will be trying to take ownership of the reference which is bad news.

> I think the way I'm spawning threads won't work with blocking?

Yep, blocking will force a task to take complete control until its completed. I would have assumed you still should have seen the internal for the task that has control to be executing but there might be another issue there.

>But the program still hanged once/twice, and even restarter hanged. No errors in output. Weird, the cpu 100% is a nightmare. now I'm testing with prints and sleep() on every continue and more things to check out what can hang in an inf loop.

For 100% its typically a sign of something spinning (i.e. as described above just looping doing nothing). Looking at the code above another place it _might_ hang is due to:
```
loop {
    while let Some(b_) = match timeout(Duration::from_secs(GET_BLOCK_TIMEOUT),stream.next()).await {
...
```
If this timeout _always_ fails (i.e. you're getting None) your going to basically spin again (with very short breaks) so that task might dominate the execution. 

> Can it hang... because of other things happening on the PC?

Typically no, unless the other thing is claiming a lock you want or something (i.e. write lock on the file system etc). However that being said it _can_ be starved (i.e. not getting a chance to run due to other things running and taking all your PC resources), highly unlikely but.

**Comment by dbelv** [comment]
> IDK how to properly debug such locks, how to get full output/stacktrace of every thread

Best bet would be to use a tool like GDB to attach to the running process and getting a stack trace for all threads.
